src/collapse/dtrace.rs

```diff
@@ -0,0 +1,214 @@
+    /// include function offset (except leafs)
+    pub includeoffset: bool,
+}
+
+/// A stack collapser for the output of dtrace `ustrace()`.
+///
+/// To construct one, either use `dtrace::Folder::default()` or create an [`Options`] and use
+/// `dtrace::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Keep track of stack string size while we consume a stack
+    stack_str_size: usize,
+
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        for _ in 0..HEADER_LINES {
+            reader.read_line(&mut line)?;
+        }
+
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(opt: Options) -> Self {
+        Self {
+            stack: VecDeque::default(),
+            occurrences: HashMap::default(),
+            opt: opt,
+            stack_str_size: 0,
+        }
+    }
+}
+
+impl Folder {
+    fn remove_offset(line: &str) -> &str {
+        let mut line = line.rsplitn(2, '+');
+        // at least one element will be returned
+        let second = line.next().unwrap();
+        if let Some(first) = line.next() {
+            first
+        } else {
+            second
+        }
+    }
+
+    // we have a stack line that shows one stack entry from the preceeding event, like:
+    //
+    //     unix`tsc_gethrtimeunscaled+0x21
+    //     genunix`gethrtime_unscaled+0xa
+    //     genunix`syscall_mstate+0x5d
+    //     unix`sys_syscall+0x10e
+    //       1
+    fn on_stack_line(&mut self, line: &str) {
+        let line = line.trim_start();
+        let frame = if self.opt.includeoffset {
+            line
+        } else {
+            Self::remove_offset(line)
+        };
+
+        let mut frame = CPP_REGEX.replace_all(frame, "$n").to_string();
+        let mut inline = false;
+
+        if frame.is_empty() {
+            frame = "-".to_string();
+        };
+
+        for func in frame.split("->") {
+            let mut func = func.replace(';', ":").trim_start_matches('L').to_string();
+            if inline {
+                func.push_str("_[i]")
+            };
+            inline = true;
+            self.stack_str_size += func.len() + 1;
+            self.stack.push_front(func);
```

Comment:
I don't _think_ this is quite right. Specifically, notice the use of `unshift` [here](https://github.com/brendangregg/FlameGraph/blob/1b1c6deede9c33c5134c920bdb7a44cc5528e9a7/stackcollapse.pl#L104). Consider something like:

```
baz
foo->bar
main
```

we want to generate

```
main;foo;bar;baz
```

The code above, unless I'm mistaken, won't do that. It'll iteratively build:

```
[baz], [foo, baz], [bar, foo, baz], [main, bar, foo, baz]
```

which results in

```
main;bar;foo;baz
```

Notice that the inlining is reversed. This is why the perf version does this strange [`push-pop-push_front`] dance, which I think we have to do here as well. We [use a `SmallVec`](https://github.com/jonhoo/inferno/blob/10a1fd7c4c891dbc6a3c6ed74541d884eb89b1e9/src/collapse/perf.rs#L278L279) for the inlined entries because _in the common case_ there isn't any inlining information (i.e., the function is just one function), and in that case `SmallVec` won't allocate anything!

Reply:
you are right, I added a test for this